src/mpi/init/init_async.c

```diff
@@ -59,22 +73,117 @@ static void progress_fn(void *data)
+                goto fn_fail;
+            }
+            thread_affinity[th_idx] = atoi(proc_id_str);
+            read_count++;
+        } else {
+            printf
+                ("ERROR: unexpected end of affinity string, expected %d numbers, read %d, affinity string %s\n",
+                 threads_per_node, read_count, affinity_to_parse);
+            mpi_errno = MPI_ERR_OTHER;
+            goto fn_fail;
+        }
+    }
+    if (read_count < threads_per_node) {
+        printf
+            ("ERROR: unexpected number of processors (specify 1 logical processor per 1 progress thread), affinity string %s\n",
+             affinity_to_parse);
+        mpi_errno = MPI_ERR_OTHER;
+        goto fn_fail;
+    }
+
+  fn_exit:
+    if (MPIR_Process.comm_world->rank == 0) {
+        for (th_idx = 0; th_idx < threads_per_node; th_idx++) {
+            MPL_DBG_MSG_FMT(MPIDI_CH4_DBG_GENERAL, VERBOSE,
+                            (MPL_DBG_FDEST, "affinity: thread %d, processor %d",
+                             th_idx, thread_affinity[th_idx]));
+        }
+    }
+    MPL_free(affinity_copy);
+    return mpi_errno;
+
+  fn_fail:
+    goto fn_exit;
+}
+
 /* called inside MPID_Init_async_thread to provide device override */
 int MPIR_Init_async_thread(void)
 {
-    int mpi_errno = MPI_SUCCESS;
+    int mpi_errno = MPI_SUCCESS, thr_err;
+    int global_rank, local_rank, local_size, threads_per_node;
+    int *thread_affinity = NULL;
     MPIR_FUNC_TERSE_STATE_DECL(MPID_STATE_MPIR_INIT_ASYNC_THREAD);
 
     MPIR_FUNC_TERSE_ENTER(MPID_STATE_MPIR_INIT_ASYNC_THREAD);
 
+    global_rank = MPIR_Process.comm_world->rank;
+    local_rank =
+        (MPIR_Process.comm_world->node_comm) ? MPIR_Process.comm_world->node_comm->rank : 0;
+    local_size =
+        (MPIR_Process.comm_world->node_comm) ? MPIR_Process.comm_world->node_comm->local_size : 1;
+    threads_per_node = local_size;
+    thread_affinity = (int *) MPL_malloc(threads_per_node * sizeof(int), MPL_MEM_OTHER);
+
+    MPL_DBG_MSG_FMT(MPIDI_CH4_DBG_GENERAL, VERBOSE,
+                    (MPL_DBG_FDEST,
+                     " global_rank %d, local_rank %d, local_size %d, threads_per_node %d",
+                     global_rank, local_rank, local_size, threads_per_node));
+
+    mpi_errno = MPIDI_parse_progress_thread_affinity(thread_affinity, threads_per_node);
+    MPIR_ERR_CHKANDJUMP(mpi_errno, mpi_errno, MPI_ERR_OTHER, "**ch4|parse_thread_affinity");
+
     int err = 0;
     MPID_Thread_create((MPID_Thread_func_t) progress_fn, NULL, &progress_thread_id, &err);
     MPIR_ERR_CHKANDJUMP1(err, mpi_errno, MPI_ERR_OTHER, "**mutex_create", "**mutex_create %s",
                          strerror(err));
 
+    MPL_thread_set_affinity(progress_thread_id, &(thread_affinity[local_rank + 1]), 1, &thr_err);
+    if (MPIR_CVAR_CH4_PROGRESS_THREAD_AFFINITY &&
+        strlen(MPIR_CVAR_CH4_PROGRESS_THREAD_AFFINITY) > 0) {
+        /* If the user did not specify the affinity, ignore affinity setting failure */
```

Comment:
Move parsing and `MPL_thread_set_affinity` both inside this condition. If the user did not set it, we shouldn't pin the threads. The code structure should reflect that.

Reply:
Moved.